presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java

```diff
@@ -86,16 +91,16 @@ public LocalDispatchQuery(
             }
         });
 
+        queryMonitor.queryCreatedEvent(stateMachine.getBasicQueryInfo(Optional.empty()));
         stateMachine.addStateChangeListener(state -> {
             if (state == QueryState.FAILED) {
-                // if query is failed and dispatching was not finished yet, marking dispatching as failed
-                submitted.set(FAILED);
-                queryExecutionFuture.cancel(true);
+                if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
+                    queryMonitor.queryImmediateFailureEvent(getBasicQueryInfo(), getFullQueryInfo().getFailureInfo());
+                }
             }
-
-            if (state.ordinal() > QueryState.DISPATCHING.ordinal()) {
-                // we went past dispatching phase; warking submitted as successful
-                submitted.set(DISPATCHED);
+            if (state.isDone()) {
+                submitted.set(null);
+                queryExecutionFuture.cancel(true);
```

Comment:
Previously `queryExecutionFuture.cancel(true);` was done only for FAILED state. Let's maybe retain that behavior, as it seems not related

Reply:
Actually a bit before my changes it was done for all finished states. Not entiryly sure why I changed that. But given fact that we are rolling back the protocol I introduced it feels natural to rollback this change too:
![image](https://user-images.githubusercontent.com/890111/101052252-b005ec00-3586-11eb-8659-89e99ff83344.png)